File: 22.括号生成/generateParenthesis.go
package problem0022

// 回溯算法是一种算法是想，递归是一种算法结构，一般回溯法多用递归来实现
func generateParenthesis(n int) []string {
	var (
		result    []string
		backTrack func(path string, left int, right int)
	)
	backTrack = func(path string, left int, right int) {
		if left == 0 && right == 0 {
			if path != "" {
				result = append(result, path)
			}
			return
		}
		// 如果左括号个数大于右括号个数，减枝，因为必定无效
		if left > right {
			return
		}
		if left > 0 {
			backTrack(path+"(", left-1, right)
		}
		if right > 0 {
			backTrack(path+")", left, right-1)
		}

	}
	backTrack("", n, n)
	return result
}

File: 22.括号生成/generateParenthesis.js
// n个左括号，n个右括号
// 回溯过程中需要保证combine合法：即 left <= right
function generateParenthesis(n) {
	const result = []
	const backTrack = (combine, left, right) => {
		if (left === 0 && right === 0) {
			result.push(combine)
			return
		}
		if (left > right) {
			return
		}
		if (left > 0) {
			backTrack(combine + '(', left - 1, right)
		}
		if (right > 0) {
			backTrack(combine + ')', left, right - 1)
		}
	}
	backTrack('', n, n)
	return result
}